File: 22.10_PS/1015_트리의_지름.py
# boj. 1167
# DFS

import sys

input = sys.stdin.readline
v = int(input())
tree = [[] for _ in range(v + 1)]
visited = [False] * (v + 1)
answer = 0
max_point = 0

for i in range(v):
    edges = list(map(int, input().split()))
    num = edges[0]
    for e in range(1, len(edges) - 1, 2):
        n, w = edges[e], edges[e + 1]
        tree[num].append((n, w))

def dfs(n, cnt):
    global answer
    global max_point

    if cnt > answer:
         max_point = n
         answer = cnt
    visited[n] = True

    for a, b in tree[n]:
        if not visited[a]:
            dfs(a, cnt + b)

    visited[n] = False

dfs(1, 0)
dfs(max_point, 0)

print(answer)

File: 22.10_PS/1015_트리의_지름.java
// boj. 1167
// BFS

import java.io.*;
import java.util.*;

public class Main {
    static ArrayList<Node>[] tree;
    static int[] weightArray;
    static boolean[] visited;

    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int v = Integer.parseInt(st.nextToken());

        tree = new ArrayList[v + 1];
        visited = new boolean[v + 1];
        weightArray = new int[v + 1];


        for (int i = 1; i < v + 1; i++) {
            st = new StringTokenizer(br.readLine());
            int num = Integer.parseInt(st.nextToken());

            tree[num] = new ArrayList<>();
            while (true) {
                int a = Integer.parseInt(st.nextToken());
                if (a == -1) {
                    break;
                }
                int b = Integer.parseInt(st.nextToken());

                tree[num].add(new Node(a, b));
            }
        }

        int nextStep = BFS(1);

        visited = new boolean[v + 1];
        weightArray = new int[v + 1];
        int answer = BFS(nextStep);

        System.out.println(weightArray[answer]);
    }

    static int BFS(int n) {
        int maxVertex = n;
        int maxValue = 0;
        Queue<Integer> queue = new LinkedList<>();
        queue.add(n);
        visited[n] = true;

        while (!queue.isEmpty()) {
            int a = queue.poll();
            for (Node t : tree[a]) {
                int v = t.vertex;
                int w = t.weight;

                if (!visited[v]) {
                    visited[v] = true;
                    queue.add(v);
                    weightArray[v] = weightArray[a] + w;
                    if (weightArray[v] > maxValue) {
                        maxValue = weightArray[v];
                        maxVertex = v;
                    }
                }
            }
        }

        return maxVertex;
    }

    static class Node {
        public int vertex;
        public int weight;

        Node(int vertex, int weight) {
            this.vertex = vertex;
            this.weight = weight;
        }
    }
}
